### PR TITLE
Add stoplight to synchronize thread starts/stops

### LINKED ISSUE #217: Synchronize thread starts

### PR DESCRIPTION
**Pull request by charmoniumQ**
This is less critical than #32. Only review if the path to #32 is blocked up.

Closes #217.

Reviewer should
- See how it works (note the `std::condition_variable` code is lifted from [C++ docs](https://en.cppreference.com/w/cpp/thread/condition_variable)).
- Verify that API of `Stoplight` is sane.
- Ensure it works on your machine.

### CODE REVIEW

File: runtime/runtime_impl.hpp
Comment (mhuzai) [inline runtime/runtime_impl.hpp:54]: Add comment saying this unblocks/starts/kicks off all threadloops?

File: runtime/runtime_impl.hpp
Comment (mhuzai) [inline runtime/runtime_impl.hpp:85]: Not a plugin.

File: runtime/runtime_impl.hpp
Comment (mhuzai) [inline runtime/runtime_impl.hpp:76]: Setting this before doing the actual stop would unblock `wait()` while the stop is in progress. Move to end of function.

File: common/stoplight.hpp
Comment (mhuzai) [inline common/stoplight.hpp:47]: Is there a reason to not use `std::lock_guard` here?

File: common/stoplight.hpp
Comment (mhuzai) [inline common/stoplight.hpp:100]: Based on this API, is my understanding correct here that stoplight is meant to be only used for one start and one stop, and cannot be used again after it has been stopped?

File: common/stoplight.hpp
Comment (mhuzai) [inline common/stoplight.hpp:84]: Comments on usage would be good.

**Review by mhuzai — COMMENTED** [review]
**Higher level decision question:** should we prohibit non-managed threads (not-threadloops) from now on because stoplight won't apply to them and they'll start working immediately? Alternatively, mandate the use of stoplight on non-managed threads? The former might be simpler.

Also, SB threads haven't been changed to use stoplight. That's fine if all the publisher threads use stoplight, but could bypass this start/stop synchronization if the publishers are non-managed threads that don't use stoplight.

File: common/threadloop.hpp
Comment (charmoniumQ) [reply common/threadloop.hpp:109]: `break` would break out of the `switch`. I want to break out of the loop which encloses the `switch` (like a `break` and then `break`). `goto` is the best way to do that in C/C++ according to [StackOverflow](https://stackoverflow.com/questions/27788326/breaking-out-of-nested-loop-c).

File: common/plugin.hpp
Comment (charmoniumQ) [reply common/plugin.hpp:28]: This does not pertain to the logging call; It pertains to `threadloop`s, since the base class wants to call `_p_iteration`, which is virtual in the child class.

File: runtime/runtime_impl.hpp
Comment (charmoniumQ) [reply runtime/runtime_impl.hpp:76]: `Stoplight::stop()` needs to happen first, and then each `threadloop::stop()`.

`Stoplight::stop()` tells the plugin threads to stop after `_p_one_iteration()`; `threadloop::stop()` joins on the plugin threads (waits for them to terminate).

File: common/stoplight.hpp
Comment (charmoniumQ) [reply common/stoplight.hpp:100]: Yeah, this is because once threads get stopped, there is no way to "unpause" them.

File: common/stoplight.hpp
Comment (charmoniumQ) [reply common/stoplight.hpp:47]: The CV wait requires a `unique_lock` so it can be re-locked, but I just changed this method to a `lock_guard` as you suggested.

**Comment by charmoniumQ** [comment]
> should we prohibit non-managed threads (not-threadloops) from now on because stoplight won't apply to them and they'll start working immediately?

Yes, they should be prohibited. There are a few places we use non-managed threads (ZED and audio_pipeline, I believe). I am switching these over in [a different PR](https://github.com/ILLIXR/ILLIXR/tree/issue-220-managed-thread-2).

> Also, SB threads haven't been changed to use stoplight. That's fine if all the publisher threads use stoplight, but could bypass this start/stop synchronization if the publishers are non-managed threads that don't use stoplight.

Yes. I wanted to keep SB independent of the rest of ILLIXR including Phonebook (although maybe I should manage that with `#ifdef`s instead).

File: common/stoplight.hpp
Comment (mhuzai) [reply common/stoplight.hpp:100]: Yeah, that makes sense. I was just wondering whether there is utility in having a reusable stoplight for, say, arbitrarily pausing the entire system and then resuming again. Could be useful for some of the simulation scenarios that we've discussed as well as for debugging/snapshots, etc. Just food for thought.

**Comment by mhuzai** [comment]
> 
> 
> > should we prohibit non-managed threads (not-threadloops) from now on because stoplight won't apply to them and they'll start working immediately?
> 
> Yes, they should be prohibited. There are a few places we use non-managed threads (ZED and audio_pipeline, I believe). I am switching these over in [a different PR](https://github.com/ILLIXR/ILLIXR/tree/issue-220-managed-thread-2).

Great. We should make this clear in the developing a plugin documentation.

> 
> > Also, SB threads haven't been changed to use stoplight. That's fine if all the publisher threads use stoplight, but could bypass this start/stop synchronization if the publishers are non-managed threads that don't use stoplight.
> 
> Yes. I wanted to keep SB independent of the rest of ILLIXR including Phonebook (although maybe I should manage that with `#ifdef`s instead).

Right, "SB-core" can be independent but perhaps "SB-ILLIXR" should use stoplight like the rest of the system. But maybe this is not necessary given that if all the plugins use stoplight, then there is no need for SB to.

File: runtime/runtime_impl.hpp
Comment (mhuzai) [reply runtime/runtime_impl.hpp:79]: Right, that makes sense. But I was talking about `runtime_impl->wait()`. As soon as you call `pb.lookup_impl<Stoplight>()->stop()`, `runtime_impl->wait()` will return, meaning that [this](https://github.com/ILLIXR/ILLIXR/blob/master/runtime/main.cpp#L98) will unblock and cause an exit possibly before the plugins have actually stopped. Or am I missing something?

File: common/plugin.hpp
Comment (mhuzai) [reply common/plugin.hpp:28]: I'm confused. Can you explain a bit more? What's the relation between `start()` and `_p_one_iteration()`?

Comment (charmoniumQ) [reply common/plugin.hpp:28]: One might think of moving that the constructor, but `plugin::plugin()` calls `threadloop::threadloop()` which calls `the_plugin::the_plugin()` at the end. So creating a thread and calling `the_plugin::_p_one_iteration()` in `threadloop::threadloop()` is not safe; `the_plugin` is not constructed until the end of `threadloop::threadloop()`.

This is why we need `plugin::start()` and `threadloop::start()`. `plugin::start()` calls `threadloop::start()` which starts a thread to call `the_plugin::_p_one_iteration()`, after `the_plugin` is constructed. This is recommended in the [ISO C++ FAQ](https://isocpp.org/wiki/faq/strange-inheritance#calling-virtuals-from-ctor-idiom) when you would need to call a virtual method from the constructor.

Comment (mhuzai) [reply common/plugin.hpp:28]: Thanks, that makes sense. But this is not clear from the current comment and needs to be documented better, so would be good to expand the comment.

File: common/threadloop.hpp
Comment (e3m3) [reply common/threadloop.hpp:109]: If we don't anticipate post-loop code blocks, we could also return directly from the switch case. Neither solution is pretty.

File: common/stoplight.hpp
Comment (charmoniumQ) [reply common/stoplight.hpp:100]: If we wanted that, I think it should go inside `Stoplight`, but in a different PR.

Comment (charmoniumQ) [reply common/stoplight.hpp:100]: That may be useful. If we wanted that, we should put it in this `Stoplight` class.

**Comment by charmoniumQ** [comment]
I will add some synchronization to Switchboard threads once #32 is merged. All else should be good. I added comments and documentation to address reviewer feedback.

File: common/runtime.hpp
Comment (e3m3) [inline common/runtime.hpp:30]:
```suggestion
		virtual ~runtime() = default;
```
I think this is slightly different semantically than an empty body (`{ }`), but I'm still not sure which should be [preferred](https://stackoverflow.com/questions/13576055/how-is-default-different-from-for-default-constructor-and-destructor). I've been switching to `= default` as I edit.

File: docs/writing_your_plugin.md
Comment (e3m3) [inline docs/writing_your_plugin.md:72]: Should we enforce this with a new derived class for non-managed types?

File: common/stoplight.hpp
Comment (e3m3) [inline common/stoplight.hpp:23]: Thoughts on using [noexcept](https://stackoverflow.com/questions/10787766/when-should-i-really-use-noexcept) for our classes?

File: common/stoplight.hpp
Comment (e3m3) [inline common/stoplight.hpp:46]: Technically this [blocks](https://en.cppreference.com/w/cpp/thread/mutex/lock) if the lock is held, but I think you are saying that this is unlikely right?

File: common/stoplight.hpp
Comment (e3m3) [inline common/stoplight.hpp:120]: :+1:

File: runtime/runtime_impl.hpp
Comment (e3m3) [reply runtime/runtime_impl.hpp:83]: Confirming: `wait_for_shutdown()` and `shutdown_done()` were added to address this right?

**Comment by e3m3** [comment]
> Ensure it works on your machine.

Clean launch and drain :+1:.

File: common/stoplight.hpp
Comment (mhuzai) [reply common/stoplight.hpp:46]: From the [condition variable documentation](https://en.cppreference.com/w/cpp/thread/condition_variable):

> Even if the shared variable is atomic, it must be modified under the mutex in order to correctly publish the modification to the waiting thread.

But this only applies to `set`, right? We should be able to make `is_set()` lock free if we can make `_m_value` atomic?

File: runtime/runtime_impl.hpp
Comment (mhuzai) [inline runtime/runtime_impl.hpp:73]: This loop is no longer required because of the blocking wait below. `wait_for_shutdown()` can only ever return if `should_stop()` is true, so this check is redundant.
The comment below might need a slight tweak without the presence of this loop.

File: common/runtime.hpp
Comment (charmoniumQ) [reply common/runtime.hpp:30]: I think `default` is preferred so that the destructor is considered "trivial." It should be trivial unless we have a reason to make it non-trivial.

File: docs/writing_your_plugin.md
Comment (charmoniumQ) [reply docs/writing_your_plugin.md:72]: Non-managed threads are quite a special case, so I don't know if a solution would generalize well.

File: common/stoplight.hpp
Comment (charmoniumQ) [reply common/stoplight.hpp:46]: Agree, will change

File: runtime/runtime_impl.hpp
Comment (charmoniumQ) [reply runtime/runtime_impl.hpp:73]: Agree, will change

File: common/stoplight.hpp
Comment (charmoniumQ) [reply common/stoplight.hpp:23]: If we later add an exception, then we would have to change the interface. I think its best to leave the option open unless there is a performance or semantic reason to do so.

Comment (e3m3) [reply common/stoplight.hpp:23]: Agreed about changing the interface. Thinking about this more, any implementation change which adds the possibility of an exception being thrown is a change to the interface, so we could make an argument that forcing these functions to `std::terminate` is more stable (i.e. introduced exceptions must be handled).

File: runtime/runtime_impl.hpp
Comment (charmoniumQ) [reply runtime/runtime_impl.hpp:83]: Yes.@mhuzai, if you are satisfied by this solution, mark conversation as resolved.

**Review by e3m3 — APPROVED** [review]

**Review by mhuzai — APPROVED** [review]
Looks great!

**Pull request merged**
